### ISSUE TITLE
pcl support

### ISSUE THREAD

**Issue by niemyjski** [grumpydev/TinyIoC#75]
We have a pcl library and we last updated in 4/22/14 and since then a lot of the new changes have broke us when we use the following directives:
# define EXPRESSIONS                         // Platform supports System.Linq.Expressions
# define COMPILED_EXPRESSIONS                // Platform supports compiling expressions

//#define APPDOMAIN_GETASSEMBLIES             // Platform supports getting all assemblies from the AppDomain object
# define UNBOUND_GENERICS_GETCONSTRUCTORS    // Platform supports GetConstructors on unbound generic types
# define GETPARAMETERS_OPEN_GENERICS         // Platform supports GetParameters on open generics
# define RESOLVE_OPEN_GENERICS               // Platform supports resolving open generics

//#define READER_WRITER_LOCK_SLIM             // Platform supports ReaderWriterLockSlim
# define TINYIOC_INTERNAL

Is anyone else seeing these issues?

**Comment by niemyjski** [comment]
We are unable to compile due to compile time errors. It looks like the logic wasn't even updated as the types on some overloads isn't even close... Also There are some exception types that are not even available..

**Comment by grumpydev** [comment]
Unfortunately "PCL support" is a bit of a misnomer, it's a bit of a car crash given how many different profiles there are - I had no real troubles recently building it into a PCL library that only supported the latest platforms (which include ReaderWriterLockSlim)

**Comment by niemyjski** [comment]
Yeah I agree. But we are using a version from april and it's been working great. We tried upgrading to the latest version and had a ton of errors. I could try doing a pull request but I'm scared that I'm breaking something. Would you be willing to pair on this, I'll drive?

Here is our project: https://github.com/exceptionless/Exceptionless and this is our pcl target:

![image](https://cloud.githubusercontent.com/assets/1020579/6415855/476bd138-be67-11e4-9c15-da9d5e323020.png)

**Comment by npodbielski** [comment]
@niemyjski 
It still does not work. Even Profile151 that suppose to work from yours latest commits is broken. I do not see any tests for PCL projects, did you test it at all?  
There are several unavailable types, methods, broken usages. 

`assembly.GetTypes();`
`BindingFlags`
`sourceType.GetMethods`
`methods.Count` is a method not property
`AppDomain`
`registrationType.IsAssignableFrom`
`MethodAccessException`
`IsAssignableFrom`
`GetConstructors`
`checkType.IsInterface` and `checkType.IsGenericType` are methods 
`type.GetGenericArguments`
`typeof(TinyIoCContainer).GetMethod`
`GetType().GetProperties()`
`registerImplementation.FindInterfaces`
Those are methods not properties
`return type.IsClass`
`return type.IsAbstract`
`return type.IsInterface`
`return type.IsPrimitive`
`return type.IsValueType`
`return type.IsGenericType`
`return type.IsGenericTypeDefinition`
`return type.BaseType`
`return type.Assembly`

I know that you are probably thinking that I am just another stupid bastard that do not know anything about your project :( , but my PULL requests with *actually working* PCL support still hangs after half of year. 

@grumpydev 
>it's a bit of a car crash given how many different profiles there are
Target most restricted PCL profile and make it work there. With this done rest would like running downhill.

**Comment by niemyjski** [comment]
I've been using this for months without issue and it's been working great. I now realize I had the link wrong to my project: https://github.com/exceptionless/Exceptionless.Net

**Comment by niemyjski** [comment]
Your running into those issues because you need to bring nuget packages available that bring in those methods / types. Please see my project.json file. https://github.com/exceptionless/Exceptionless.Net/blob/v4.0.2/src/Exceptionless/project.json

**Comment by npodbielski** [comment]
Thanks for quick response.

You do realize that, even if this works in your project it do not have to **actually work** in others? 
PCL is not the same as NETCore.
Anyway if the answer to this bugs is: install some nuget packages there is something fundamentally wrong with this whole project.
>Welcome to TinyIoC - an easy to use, hassle free, Inversion of Control Container. 

If you have to do some extra work to make it... um work... on some platform and this is never mentioned in documentation at all... I would not call it neither "easy", nor "hassle free".
>TinyIoC has been designed to fulfil a single key requirement - to lower the "level of entry" for using an IoC container

If it was designed to work as a single file, why you do not follow this "single key requirement" while supposedly adding PCL support. Did you created another Nuget package just for PCL, to make those dependencies visible? I can't find it.
Did you edited Readme to **mark** that some **extra** steps needs to be performed to make TinyIoC work in PCL projects?

I am sad, that your response can be summarized as "It is working on **my** machine". :(

**Comment by npodbielski** [comment]
Just to be sure, I tried to find and install Nuget packages you have listed in your .json file.
And as I expected it **did exactly nothing**. After all, if maybe there is some package with `AppDomain` class for PCL (which I highly doubt it), it is impossible to change member type from method to property just by adding some reference (and there are errors like that, as I stated in my first comment and you read it whole, right?). 
This is how my project references looks like after adding bunch of nuget packages (not all because, somel are not present in feed and some are incompatible with PCL profile 151).
![obraz](https://cloud.githubusercontent.com/assets/796782/23496308/b8727dfc-ff1d-11e6-9a18-4e546b07a00e.png)
This is **probably** because all of those useless packages looks like this:
![obraz](https://cloud.githubusercontent.com/assets/796782/23496354/0218a652-ff1e-11e6-8a14-85b7721d454b.png)
